crates/llms/src/model2vec.rs

```diff
@@ -0,0 +1,206 @@
+    /// * `embed_batch_size` - Batch size for embedding operations
+    /// * `embed_custom_batch_size` - Custom batch size override
+    ///
+    /// # Errors
+    /// Returns an error if:
+    /// - The model cannot be loaded or initialized
+    /// - Invalid parameters are provided
+    /// - Network issues occur when downloading the model
+    /// - Authentication fails with the provided HF token
+    pub fn from_params(
+        name: &str,
+        hf_token: Option<&str>,
+        normalize: Option<bool>,
+        subfolder: Option<&str>,
+        parallelism: Option<usize>,
+        embed_max_token_length: Option<usize>,
+        embed_custom_batch_size: Option<usize>,
+    ) -> Result<Self, super::embeddings::Error> {
+        let model = StaticModel::from_pretrained(name, hf_token, normalize, subfolder)
+            .map_err(|e| FailedToInstantiateEmbeddingModel { source: e.into() })?;
+
+        let model2vec = Self {
+            name: name.to_string(),
+            model,
+            normalize,
+            parallelism,
+            embed_max_token_length,
+            embed_custom_batch_size,
+        };
+
+        tracing::trace!("Model2Vec::from_params: {model2vec:?}");
+
+        Ok(model2vec)
+    }
+}
+
+impl Debug for Model2Vec {
+    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
+        let Self {
+            name,
+            normalize,
+            parallelism,
+            embed_max_token_length,
+            embed_custom_batch_size,
+            ..
+        } = self;
+        write!(
+            f,
+            "Model2Vec: {name}, normalize: {normalize:?}, parallelism: {parallelism:?}, embed_max_token_length: {embed_max_token_length:?}, embed_custom_batch_size: {embed_custom_batch_size:?}"
+        )
+    }
+}
+
+#[async_trait]
+impl Embed for Model2Vec {
+    async fn embed(
+        &self,
+        input: EmbeddingInput,
+    ) -> Result<Vec<Vec<f32>>, super::embeddings::Error> {
+        self.embed_sync(input)
+    }
+
+    fn embed_sync(&self, input: EmbeddingInput) -> Result<Vec<Vec<f32>>, super::embeddings::Error> {
+        let embedding_input = match input {
+            EmbeddingInput::String(s) => vec![s],
+            EmbeddingInput::StringArray(sentences) => sentences,
+            _ => {
+                return Err(UnsupportedEmbeddingInput {
+                    model: self.name.clone(),
+                    message: "Model2Vec models only support strings or vectors of strings"
+                        .to_string(),
+                });
+            }
+        };
+
+        Ok(self.model.encode_with_args(
+            &embedding_input,
+            self.embed_max_token_length,
+            self.embed_custom_batch_size.unwrap_or(1024),
+        ))
+    }
+
+    fn supports_sync_embeddings(&self) -> bool {
+        true
+    }
+
+    fn parallelism(&self) -> Option<usize> {
+        self.parallelism.or(Some(0))
```

Comment:
Why is `Some(0)` and not `Some(1)` the correct default?

Reply:
OH! Yeah I'm cool with `supports_sync_embeddings` gating opt-in to parallelism. Which means we can use `None`. Changing!!